util/component/mux.go

```diff
@@ -0,0 +1,186 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package component
+
+import (
+	"context"
+	"flag"
+	"fmt"
+
+	utilflag "github.com/kubewharf/podseidon/util/flag"
+	"github.com/kubewharf/podseidon/util/iter"
+	"github.com/kubewharf/podseidon/util/optional"
+	"github.com/kubewharf/podseidon/util/util"
+)
+
+// Provides an implementation that implements `Interface`,
+// declaring a component for the implementation and a component for the interface.
+// If the interface component is declared multiple times,
+// an implementation is selected based on the CLI options provided by the user.
+//
+// The same `implName` must not be declared twice for the same `muxName`.
+// All calls to `DeclareMuxImpl` with the same `muxName` must pass the same `Interface`.
+//
+// All options are prefixed by `{muxName}-{implName}` instead of `{implName}`.
+// Implementation components that did not get selected by the user
+// and components only requested by such components
+// do not get `init`ed, and hence their lifecycles are never invoked.
+//
+// The first implementation that got declared is always the default option.
+func DeclareMuxImpl[Args any, Options any, Deps any, State any, Interface any](
+	muxName string,
+	implNameFn func(Args) string,
+	optionsFn func(Args, *flag.FlagSet) Options,
+	depsFn func(Args, *DepRequests) Deps,
+	init func(context.Context, Args, Options, Deps) (*State, error),
+	lifecycle Lifecycle[Args, Options, Deps, State],
+	api func(*Data[Args, Options, Deps, State]) Interface,
+) func(Args) func(*DepRequests) {
+	return func(implArgs Args) func(*DepRequests) {
+		implName := implNameFn(implArgs)
+
+		implComp := Declare(
+			func(Args) string { return fmt.Sprintf("%s-%s", muxName, implName) },
+			optionsFn,
+			depsFn,
+			init,
+			lifecycle,
+			func(d *Data[Args, Options, Deps, State]) MuxImpl[Interface] {
+				return MuxImpl[Interface]{
+					getImpl: func() Interface { return api(d) },
+				}
+			},
+		)(implArgs)
+
+		apiComp := declareApiComp[Interface](
+			muxName,
+		).WithMergeFn(func(prevArgs *muxArgs[Interface], prevDeps *muxDeps[Interface], reqs *DepRequests) {
+			prevArgs.impls[implName] = implComp
+			prevDeps.impls[implName] = DepPtr(reqs, implComp)
+
+			if prevArgs.defaultOption.IsNone() {
+				prevArgs.defaultOption = optional.Some(implName)
+			}
+		})(
+			muxArgs[Interface]{
+				description:   optional.None[string](), // to be populated by DeclareMuxInterface
+				impls:         map[string]Declared[MuxImpl[Interface]]{implName: implComp},
+				defaultOption: optional.Some(implName),
+			},
+		)
+
+		return RequireDep(apiComp)
+	}
+}
+
+// Declares a mux interface to request a dependency to the selected mux implementation.
+// See [DeclareMuxImpl] for details.
+// The returned value can be stored in a global to serve as a function like [Declare].
+func ProvideMux[Interface any](muxName string, description string) func() Declared[Interface] {
```

Comment:
How about specifying the default provider explicitly when defining the `provideMux`?

Reply:
wouldn't this mean the interface has to be coupled with at least one of its implementations?

Reply:
I'll make it specified by the main `cmd.Run` during startup instead.